Maker-Agents/Governance-Module/src/hat.fetcher.ts

```diff
@@ -1,30 +1,35 @@
 import abi from "./abi";
 import { providers, Contract, BigNumber } from "ethers";
+import AddressFetcher from "./address.fetcher";
 
 export default class HatFetcher {
   private hat: string;
   private block: number;
   private approvals: BigNumber;
-  private chiefContract: Contract;
+  chiefFetcher: AddressFetcher;
+  provider: providers.JsonRpcProvider;
 
-  public constructor(chief: string, provider: providers.JsonRpcProvider) {
+  public constructor(chiefFetcher: AddressFetcher, provider: providers.JsonRpcProvider) {
     this.block = -1;
     this.hat = "None";
     this.approvals = BigNumber.from(-1);
-    this.chiefContract = new Contract(chief, abi.CHIEF, provider);
+    this.chiefFetcher = chiefFetcher;
+    this.provider = provider;
   }
 
   public async getHat(block: number) {
     if (this.block !== block) {
       this.block = block;
       this.approvals = BigNumber.from(-1);
-      this.hat = await this.chiefContract.hat({ blockTag: block });
+      const contract = new Contract(this.chiefFetcher.chiefAddress, abi.CHIEF, this.provider);
+      this.hat = await contract.hat({ blockTag: block });
     }
     return this.hat;
   }
 
   public async getHatApprovals(block: number) {
-    if (this.approvals.eq(-1)) this.approvals = await this.chiefContract.approvals(this.hat, { blockTag: block });
+    const contract = new Contract(this.chiefFetcher.chiefAddress, abi.CHIEF, this.provider);
```

Comment:
Is it possible to have this file be where you don't need to create a contract instance every time either of these functions are called? 

I'm not sure if having `this.contract = new Contract(this.chiefFetcher.chiefAddress, abi.CHIEF, this.provider);` in the `constructor` would work, since `this.chiefFetcher.chiefAddress` could potentially have a value of `""`.

Reply:
there is no way because that contract can change, so initializing the contract at the beginning is not an option